### PR TITLE
[WB-7243] [WB-5837] Fix base url to allow trailing /

### PR DESCRIPTION
**Pull request by kptkin**
https://wandb.atlassian.net/browse/WB-7243
https://wandb.atlassian.net/browse/WB-5837

Description
-----------

What does the PR do?

Removes trailing `/`s from a base url passed to the Public API. This avoids a wrong url when the Public API's Client is trying to hit the GraphQL endpoint.

Also Fixed the Environment variable trailing slash which solves the settings file and os.environ.
with settings file:
```
[default]
base_url = https://api.wandb.ai//
``` 
**Before the Env fix (settings):**
image
image

**After the Env fix (settings):**
image

**Before the Env fix:**
image
image

**After the Env fix:**
image

Testing
-------

How was this PR tested?

Pytest that checks that url is being sanitized and manually validate that the issue is resolved.

Release Notes
-------------

Below, please enter user-facing release notes as one or more bullet points.
If your change is not user-visible, write `NO RELEASE NOTES` instead, with no bullet points.

------------- BEGIN RELEASE NOTES ------------------

NO RELEASE NOTES

------------- END RELEASE NOTES --------------------

### CODE REVIEW

**Review by adrnswanberg — APPROVED** [review]

**Review by vanpelt — COMMENTED** [review]
Ideally we wouldn't be doing this in a one-off way everywhere.  The "right" way is to pipe everything through wandb.Settings but that's a larger exercise.  One thing I would check is what happens if you do:

```
wandb login --host=https://foo.bar.com/
```

**Comment by kptkin** [comment]
> Ideally we wouldn't be doing this in a one-off way everywhere. The "right" way is to pipe everything through wandb.Settings but that's a larger exercise. One thing I would check is what happens if you do:
> 
> ```
> wandb login --host=https://foo.bar.com/
> ```

I agree! That's what I was originally tring to do, but it wasn't a quick fix, since we are using settings all over and it's not properly encapsulated, so will require refactoring how we use settings everywhere.

I tried 
```
wandb login --host=https://foo.bar.com/
```
and works without any fixes:


> wandb login --host=https://foo.bar.com/

(works without the current fixes)
image

**Review by vanpelt — APPROVED** [review]

**Review by raubitsj — APPROVED** [review]
Thanks @kptkin ,  unified settings i think isnt too far away, but this is great for now.

**Comment by kptkin** [comment]
> Thanks @kptkin , unified settings i think isnt too far away, but this is great for now.

I will open a ticket for this in Jira. Just to track it.

File: wandb/apis/public.py [hunk 1]
```diff
@@ -287,15 +287,22 @@ class Api(object):
...
-            self.settings["entity"] = overrides["username"]
+            overrides["entity"] = overrides["username"]
+
+        _api = InternalApi()
+        for key, value in overrides.items():
+            _api.set_setting(key, value)
+
+        self.settings = _api.settings()
+
+        if self.api_key is None:
```
Comment (kptkin) [inline wandb/apis/public.py:303]: @vanpelt @raubitsj I think this is the correct order. `api_key` depends on `self.settings['base_url']` so in the old way we would check for the old settings and then update it. Seems wrong to me, I assume the `base_url` should be the same for login and for client

**Pull request merged**